충남대 FE 정지원 - 1단계 선물하기 메인 API 구현하기

jjw5655 opened the pull request:
1단계 구현입니다. 늦어서 죄송합니다...

마지막 커밋이 어쩌다보니 덩치가 커졌습니다. 앞으로는 더 분리하려고 노력해볼게요.

jjw5655 committed:
chore: Axios 패키지 추가

- HTTP 요청을 위한 Axios 라이브러리 설치
  - npm install axios로 설치
- API 클라이언트 구현을 위한 의존성 추가

jjw5655 committed:
feat: API 클라이언트 및 API 호출 함수 구현

- Axios 기반 API 클라이언트 설정
- .env 파일에 API URL 환경변수 설정 (VITE_API_BASE_URL=http://localhost:3000)
  - .gitignore에 의해 git에 올라가지는 않음
- 테마 및 상품 API 타입 정의
- 테마 목록 조회 API 함수 구현
- 실시간 급상승 선물 랭킹 조회 API 함수 구현

jjw5655 committed:
feat: 데이터 페칭 커스텀 훅 구현

- useFetch: 기본 데이터 페칭 로직 및 상태 관리
- useThemes: 테마 목록 조회 훅
- useRankingProducts: 실시간 급상승 선물 랭킹 조회 훅
- hooks/index.ts 업데이트

jjw5655 committed:
fix: API 응답 구조에 맞게 타입 및 컴포넌트 수정

- API 응답 { data: [...] } 구조에 맞게 타입 정의 수정
- ThemeResponse 인터페이스의 themes -> data로 변경
- CategorySection 컴포넌트에서 데이터 처리 로직 수정
- HomePage에서 mock 데이터 대신 API를 통해 데이터를 가져오는 방식으로 변경

jjw5655 committed:
feat: 실시간 급상승 선물랭킹 및 상품 상세 페이지 API 연동 구현

- 실시간 급상승 선물랭킹 섹션 API 연동
  * API 응답 구조에 맞게 상품 관련 타입 정의 수정
  * ProductCard, ProductGrid 컴포넌트 API 구조에 맞게 수정
  * 로딩 상태 및 에러 상태 처리 추가
  * 상품 목록이 없는 경우 안내 메시지 표시
  * useRankingProducts 훅을 통한 데이터 페칭

- 상품 상세 정보 API 연동
  * getProductById API 함수 추가
  * useProduct 커스텀 훅 구현
  * OrderPage를 mock 데이터 대신 API 사용하도록 수정
  * 상품 클릭 시 실제 상품 ID로 주문 페이지 이동
  * 로딩 및 에러 상태 처리 추가

- 데이터 흐름 개선
  * RankingSection에서 실제 상품 ID를 사용하여 OrderPage로 이동
  * API 응답 구조 통일 (ProductResponse 타입 추가)
  * rankingUtils에서 Product 타입 import 경로 수정
  * 전체적인 사용자 경험 향상

jjw5655 committed:
docs: 미션 3 - 1단계 과제 요구사항으로 README.md 업데이트

- 선물하기 메인 API 구현하기 과제 요구사항 추가
- 선물 테마 섹션 API 구현 요구사항 명시
- 실시간 급상승 선물랭킹 섹션 API 구현 요구사항 명시
- 로딩 화면 및 에러 처리 요구사항 포함

ed-jinyoung-park commented on src/hooks/useFetch.ts:
```diff
@@ -0,0 +1,53 @@
+
+  useEffect(() => {
+    fetchData();
+  }, [url, JSON.stringify(options)]); // options 객체 변경 감지를 위해 stringify
```
stringify하지 않아도 options 객체의 참조로도 변경을 감지할 수 있지 않은가요?

ed-jinyoung-park commented on src/components/ranking/RankingSection.tsx:
```diff
@@ -1,20 +1,67 @@
 type TargetType = (typeof targetOptions)[number]['value'];
 type RankType = (typeof rankOptions)[number]['value'];
 
+// 로딩 스켈레톤 컴포넌트
```
로딩 스켈레톤 컴포넌트가 CategorySection과 동일하게 중복되어 있는 것 같아서 공통화가 가능하다면 공용 컴포넌트로 빼보시는 것도 좋을 것 같습니다! 

별개로 스켈레톤용 style 컴포넌트가 RankingSection의 관심사에서는 중요하지 않은 부분인데 코드가 다소 길어 가독성을 해치는 것 같아서 공통화가 안되더라고 각각 별도 컴포넌트로 빼보는 것도 추천드립니다!

jjw5655 replied:
공용으로 빼진 않고 각각에 대해 별도의 skeleton파일로 분리했습니다!

ed-jinyoung-park commented on src/components/ranking/RankingSection.tsx:
```diff
@@ -38,31 +85,15 @@ const RankingSection = () => {
-      ); // replace로 히스토리에 남기지 않음
-    }
-  }, [targetParam, rankParam, targetType, rankType, setSearchParams]);
+  // API filter 매핑 - target과 rank를 합쳐서 필터로 변환
+  const getFilterFromParams = (): FilterOption => {
+    // 현재는 API가 단순 필터만 받으므로 'all'로 고정
+    // 향후 API에서 target/rank 조합을 지원하면 이 부분 확장
+    return 'all';
+  };
```
[수정요청]

[API 명세](https://github.com/next-step/react-gift-mock-server/blob/main/API_DOCS.md#46-%EC%83%81%ED%92%88-%EB%9E%AD%ED%82%B9-%EC%A1%B0%ED%9A%8C)를 보시면 target과 rank를 합친 필터가 아닌 target과 rank 각각 별도 파라미터로 받고 있는데요. 화면에서 선택한 target과 rank를 params로 받아 api를 호출하도록 수정이 필요할 것 같습니다!

jjw5655 replied:
완전 잘못되어있었네요. 수정했습니다!

ed-jinyoung-park commented on src/components/ranking/RankingSection.tsx:
```diff
@@ -98,15 +145,23 @@ const RankingSection = () => {
-        showMore={showMore}
-        onProductClick={handleProductClick}
-      />
+      {isLoading ? (
+        renderLoadingSkeleton()
+      ) : error ? null : ( // 에러시 아무것도 표시하지 않음
```
삼항연산이 이중으로 들어가있어 코드를 읽을때 혼동을 줄 수 있을 것 같아요.
if문이 많이 들어가더라도 읽기 쉬운 아래와 같은 방법을 추천드립니다!


```suggestion
if(isLoading){
   return renderLoadingSkeleton();
}

if(error){
  return null;
}
```

jjw5655 replied:
수정하였습니다.
이 코드처럼 더 짧게 작성할 수 있는 코드를 가독성을 위해 어느정도로 펼쳐서 작성해야 할지 고민될 때가 많은데, 협업을 고려하면 그냥 다 펼쳐서 if문으로 쉽게 적는게 더 나을거같네요. 제안 감사합니다!

ed-jinyoung-park commented on src/components/ranking/RankingSection.tsx:
```diff
@@ -98,15 +145,23 @@ const RankingSection = () => {
-      <MoreButton onClick={() => setShowMore(!showMore)}>
-        {showMore ? '접기' : '더보기'}
-      </MoreButton>
+          {data?.data && data.data.length > 6 && (
```
[수정요청]

api 응답값에 따라 showMore가 true인데도 data.length가 6보다 아래인 경우가 있을 수 있습니다. 이부분은 기존처럼 showMore로 분기처리 하는 것으로 수정하는게 적절해보입니다!
```suggestion
          {data?.data && showMore && (
```

jjw5655 replied:
showMore의 초기값이 false(더보기)이고, 누를때마다 true(접기)와 false(더보기)가 바뀌는 상태입니다. 제안해주신 방법으로 수정하니 초기값이 false여서 버튼이 아예 보이지 않게 되는데, 어디가 문제이고 어떻게 수정해야할지 조금 더 자세히 설명해주실 수 있나요?

ed-jinyoung-park replied:
이부분은 제가 헷갈렸네요 말씀하신 대로 기존 로직이 맞네요 죄송합니다 ㅠ

ed-jinyoung-park commented on src/api/products.ts:
```diff
@@ -0,0 +1,36 @@
+ * 실시간 급상승 선물 랭킹을 조회합니다.
+ * @param filter - 필터 옵션 (all, price-asc, price-desc, popular)
+ */
+export const getRankingProducts = async (
+  filter: FilterOption = 'all'
+): Promise<RankingResponse> => {
+  const response = await apiClient.get<RankingResponse>(
+    '/api/products/ranking',
+    {
+      params: { filter },
+    }
+  );
+  return response.data;
+};
```
[수정요청]

[API명세](https://github.com/next-step/react-gift-mock-server/blob/main/API_DOCS.md#46-%EC%83%81%ED%92%88-%EB%9E%AD%ED%82%B9-%EC%A1%B0%ED%9A%8C) 와 params 스펙이 다른 것 같아요. 명세에 맞게 수정 요청드립니다!

ed-jinyoung-park commented on src/hooks/index.ts:
```diff
@@ -1,5 +1,9 @@
-export { useForm } from './useForm';
 export { useLocalStorageState } from './useLocalStorageState';
 export { useBodyScrollLock } from './useBodyScrollLock';
+export { default as useFetch } from './useFetch';
+export { default as useThemes } from './useThemes';
+export { default as useRankingProducts } from './useRankingProducts';
+export { default as useProduct } from './useProduct';
```
다른 부분은 named export를 하고 있는데 fetch 관련 hook들만 default export를 하신 이유가 있을까요? 일관성을 위해 하나로 통일하는 것이 좋아보입니다!

jjw5655 replied:
통일시켰습니다!

ed-jinyoung-park commented on src/hooks/useProduct.ts:
```diff
@@ -0,0 +1,42 @@
+ * 특정 상품의 상세 정보를 가져오는 커스텀 훅
+ * @param productId - 상품 ID
+ */
+function useProduct(productId: string | number | undefined) {
+  const [product, setProduct] = useState<Product | null>(null);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<Error | null>(null);
+
+  useEffect(() => {
+    if (!productId) {
+      setError(new Error('상품 ID가 필요합니다.'));
+      setIsLoading(false);
+      return;
+    }
+
+    const fetchProduct = async () => {
+      try {
+        setIsLoading(true);
+        const productResponse = await getProductById(productId);
+        setProduct(productResponse.data); // API 응답에서 .data 필드 사용
+        setError(null);
+      } catch (err) {
+        console.error('상품 정보 가져오기 실패:', err);
+        setError(err as Error);
+        setProduct(null);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    fetchProduct();
+  }, [productId]);
+
+  return { product, isLoading, error };
```
이 부분은 1단계 구현사항은 아니긴 하지만 useTheme과 동일하게 useFetch를 사용해서 fetchState를 간결하게 관리해보시는 것을 추천드립니다!

jjw5655 replied:
구현해둔 useFetch를 사용하도록 수정했습니다.

ed-jinyoung-park commented on src/hooks/useRankingProducts.ts:
```diff
@@ -0,0 +1,53 @@
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<Error | null>(null);
+
+  const fetchRankingProducts = async (currentFilter: FilterOption) => {
+    try {
+      setIsLoading(true);
+      const response = await axios.get<RankingResponse>(
+        'http://localhost:3000/api/products/ranking',
+        {
+          params: { filter: currentFilter },
+        }
+      );
+      setData(response.data);
+      setError(null);
+    } catch (err) {
+      setError(err as Error);
+      setData(null);
+      console.error('Error fetching ranking products:', err);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    fetchRankingProducts(filter);
+  }, [filter]);
+
+  const changeFilter = (newFilter: FilterOption) => {
+    setFilter(newFilter);
+  };
+
+  return {
+    data,
+    isLoading,
+    error,
+    filter,
+    changeFilter,
+    refetch: () => fetchRankingProducts(filter),
+  };
+}
```
[수정요청] 이부분도 API명세와 다르게 구현되어 있는 것 같아요. 수정 부탁드릴게요! 수정하실 때 useFetch를 만드신 만큼 useFetch를 사용해서 비동기 데이터 로직에 대해 data, isLoading, error 상태를 일관되게 관리해보시는 것을 권장드립니다.

jjw5655 replied:
얘도 마찬가지로 useFetch를 사용하도록 수정했습니다.

ed-jinyoung-park reviewed (changes requested):
지원님 1단계 고생하셨습니다!

커밋의 단위는 작업하다 보면 저도 가끔 커지기 마련입니다. git을 다루는데 익숙해지고 커밋 단위를 쪼개는 연습을 하시다 보면 자연스레 좋아지실 것 같아요!

살펴보니 요구사항 중 하나인 [상품 랭킹 조회 API 스펙](https://github.com/next-step/react-gift-mock-server/blob/main/API_DOCS.md#46-%EC%83%81%ED%92%88-%EB%9E%AD%ED%82%B9-%EC%A1%B0%ED%9A%8C)과 다르게 구현된 부분이 보이네요. [수정요청]으로 코멘트를 달아두었으니 확인해보시고 수정 부탁드립니다!

jjw5655 committed:
perf: useFetch 훅의 성능 최적화 - JSON.stringify 제거 및 메모이제이션 적용

- JSON.stringify() 제거로 불필요한 연산 최소화
- useMemo로 options 객체 참조 안정화
- useCallback으로 fetchData 및 refetch 함수 메모이제이션
- 의존성 배열을 객체 참조 기반으로 변경하여 성능 향상

jjw5655 committed:
refactor: RankingSection 스켈레톤 컴포넌트 분리

- RankingSkeleton 컴포넌트로 스켈레톤 UI 분리
- RankingSection에서 스켈레톤 스타일 코드 제거
- 관심사 분리로 가독성 향상 및 재사용성 증대

jjw5655 committed:
refactor: CategorySection 스켈레톤 컴포넌트 분리

- CategorySkeleton 컴포넌트로 스켈레톤 UI 분리
- CategorySection에서 스켈레톤 스타일 코드 제거
- 관심사 분리로 가독성 향상 및 재사용성 증대

jjw5655 committed:
chore: API 명세 파일 추가 및 gitignore 설정

- API_DOCS.md 파일 추가 (참고용)
- .gitignore에 API_DOCS.md 추가하여 git 추적 제외

jjw5655 committed:
fix: 랭킹 API 호출 방식을 API 명세에 맞게 수정

- FilterOption 타입 제거 및 TargetType, RankType 타입 추가
- getRankingProducts API 호출 시 targetType, rankType 각각 전달
- useRankingProducts 훅에서 targetType, rankType 분리 처리
- RankingSection에서 getFilterFromParams 함수 제거
- API 호출: /api/products/ranking?targetType=ALL&rankType=MANY_WISH

jjw5655 committed:
refactor: RankingSection 조건부 렌더링 개선

- 삼항 연산자 중첩을 if문으로 변경하여 가독성 향상
- 복잡한 화살표 함수를 renderContent 함수로 분리
- 조기 반환 패턴 적용으로 코드 흐름 명확화
- 로딩, 에러, 정상 상태별 명확한 분리

jjw5655 committed:
refactor: hooks import/export 방식 named export로 통일 및 관련 에러 수정

- useFetch, useThemes, useRankingProducts, useProduct를 named export로 변경
- hooks/index.ts 및 각 훅 파일 export 방식 일관성 유지
- 프로젝트 내 import 구문을 모두 중괄호(named) 방식으로 변경하여 타입 에러 해결

jjw5655 committed:
fix: useProduct 훅에 productId undefined 타입 처리

- productId가 undefined일 때 빈 문자열로 대체하여 타입 에러 해결
- OrderPage에서 useProduct 호출 시 타입 안전성 강화

jjw5655 committed:
refactor: useRankingProducts 훅 useFetch 기반 리팩토링 및 절대경로 적용

- useFetch를 활용해 data, isLoading, error 상태를 일관되게 관리
- API 명세에 맞는 URL 및 파라미터 사용
- 불필요한 상태 및 effect, 직접 fetch 로직 제거로 코드 간결화
- useRankingProducts에서 url을 baseURL을 포함한 절대경로로 생성
- axios 기본 인스턴스 사용 시 항상 백엔드로 요청이 가도록 수정
- 상품 랭킹 등 API 정상 동작 보장

jjw5655 committed:
refactor: useFetch 기반으로 랭킹/상품상세 훅 리팩토링 및 절대경로 적용

- useRankingProducts, useProduct 훅을 useFetch 기반으로 리팩토링하여 data, isLoading, error 상태를 일관되게 관리
- API 명세에 맞는 URL 및 파라미터 사용
- 불필요한 상태 및 effect, 직접 fetch 로직 제거로 코드 간결화
- useRankingProducts, useProduct에서 url을 baseURL을 포함한 절대경로로 생성
- axios 기본 인스턴스 사용 시 항상 백엔드로 요청이 가도록 수정
- 상품 랭킹, 상품 상세 등 API 정상 동작 보장

jjw5655 commented:
원래 금일 7.17(목) 18:00 까지 2단계 제출을 해야하는데 조금 늦어질 것 같습니다...
지금 진행중인 1단계 머지해주시면 최대한 빨리 제출하겠습니다!

jjw5655 commented:
안녕하세요 박진영 멘토님! 리뷰해주신 내용들을 반영하였습니다.

원래 금일 7.17(목) 18:00 까지 2단계 제출을 해야하는데 조금 늦어질 것 같습니다...
지금 진행중인 1단계 pr 머지해주시면 최대한 빨리 제출하겠습니다!

ed-jinyoung-park commented on src/constants/rankingOptions.ts:
```diff
@@ -1,12 +1,14 @@
-  { value: 'FEMALE', label: '여성이' },
-  { value: 'MALE', label: '남성이' },
-  { value: 'TEEN', label: '청소년이' },
+  { value: 'ALL' as TargetType, label: '전체' },
+  { value: 'FEMALE' as TargetType, label: '여성이' },
+  { value: 'MALE' as TargetType, label: '남성이' },
+  { value: 'TEEN' as TargetType, label: '청소년이' },
```
as assertion을 사용하기 보다는 targetOptions라는 배열에 대한 type 정의를 해주시면 타입추론이 더 용이해질 것 같아요!
```ts
const targetOptions : {value: TargetType, label: string}[]
```

ed-jinyoung-park commented on src/constants/rankingOptions.ts:
```diff
@@ -1,12 +1,14 @@
-  { value: 'MANY_WISH', label: '받고 싶어한' },
-  { value: 'MANY_RECEIVE', label: '많이 선물한' },
-  { value: 'MANY_WISH_RECEIVE', label: '위시로 받은' },
+  { value: 'MANY_WISH' as RankType, label: '받고 싶어한' },
+  { value: 'MANY_RECEIVE' as RankType, label: '많이 선물한' },
+  { value: 'MANY_WISH_RECEIVE' as RankType, label: '위시로 받은' },
```
여기도 위와 동일합니다.

ed-jinyoung-park commented on src/components/ranking/RankingSection.tsx:
```diff
@@ -30,39 +25,16 @@ const RankingSection = () => {
     targetParam,
     targetValidValues,
-    'ALL' as TargetType
+    'ALL' as (typeof targetOptions)[number]['value']
```
기존에 TargetType으로도 타입 에러가 없을텐데 바꾸신 이유가 있을까요?

ed-jinyoung-park reviewed (approved):
지원님 API 스펙 수정과 리팩토링까지 수정사항과 코멘트들을 잘 반영해주셨네요.

고생하셨습니다 머지하겠습니다!

type 관련 추가 코멘트만 확인해주세요!
